proposals/init.md

```diff
@@ -0,0 +1,582 @@
+object's immutable properties can participate in object initialzers and thus
+removes the need for all constructor boiler plate in the type. The `Point`
+type is now simply:
+
+```cs
+struct Point
+{
+    public int X { get; init; }
+    public int Y { get; init; }
+}
+```
+
+The consumer can then use object initializers to create the object
+
+```cs
+var p = new Point() { X = 42, Y = 13 };
+```
+
+## Detailed Design
+
+### init members
+An init only property is declared by using the `init` accessor in place of the 
+`set` accessor:
+
+```cs
+class Student
+{
+    public string FirstName { get; init; }
+    public string LastName { get; init; }
+}
+```
+
+An instance property containing an `init` accessor is considered settable in
+the following circumstances:
+
+- During an object initializer
+- Inside an instance constructor of the containing or derived type
+- Inside the `init` accessor of any property on `this`
+
+The times above in which the `init` accessors are settable are collectively
+referred to in this document as the construction phase of the object.
+
+This means the `Student` class can be used in the following ways:
+
+```cs
+var s = new Student()
+{
+    FirstName = "Jared",
+    LastName = "Parosns",
+};
+s.LastName = "Parsons"; // Error: LastName is not settable
+```
+
+The rules around when `init` accessors are settable extend across type
+hierarchies. If the member is accessible and the object is known to be in the
+construction phase then the member is settable. That specifically allows for
+the following:
+
+```cs
+class Base
+{
+    public bool Value { get; init; }
+}
+
+class Derived : Base
+{
+    Derived()
+    {
+        // Not allowed with get only properties but allowed with init
+        Value = true;
+    }
+}
+
+class Consumption
+{
+    void Example()
+    {
+        var d = new Derived() { Value = true; };
+    }
+}
+
+```
+
+At the point a `init` accessor is invoked the instance is known to be 
+in the open construction phase. Hence an `init` accessor is allowed to take 
+the following actions in addition to what a normal `set` accessor cand do:
+
+1. Call other `init` accessors available through `this`
+1. Assign `readonly` fields declared on the same type
+
+```cs
+class Complex
+{
+    readonly int Field1;
+    int Field2;
+    int Prop1 { get; init ; }
+    int Prop2
+    {
+        get => 42;
+        set
+        {
+            Field1 = 13; // okay
+            Field2 = 13; // okay
+            Prop1 = 13; // okay
+        }
+    }
+}
+```
+
+The ability to assign `readonly` fields from an `init` accessor is limited to 
+those fields declared on the same type as the accessor. It cannot be used to 
+assign `readonly` fields in a base type. This rule ensures that type authors
+remain in control over the mutability behavior of their type. Developers who do
+not wish to utilize `init` cannot be impacted from other types choosing to
+do so:
+
+```cs
+class Base
+{
+    internal readonly int Field;
```

Comment:
Maybe make the field lower-case?

Reply:
For design docs I usually leave all members as PascalCased. Feel like it makes the document easier to read.

Reply:
Alright, I thought the exact opposite while reading this but I’m fine with it if this is like a design docs convention 😀